src/Common/LRUCachePolicy.h

```diff
@@ -0,0 +1,186 @@
+    LRUCachePolicy(OnWeightLossFunction on_weight_loss_function_, Args... args) : LRUCachePolicy(args...)
+    {
+        Base::on_weight_loss_function = on_weight_loss_function_;
+    }
+
+    size_t weight([[maybe_unused]] std::lock_guard<std::mutex> & cache_lock) const override
+    {
+        return current_size;
+    }
+
+    size_t count([[maybe_unused]] std::lock_guard<std::mutex> & cache_lock) const override
+    {
+        return cells.size();
+    }
+
+    size_t maxSize() const override
+    {
+        return max_size;
+    }
+
+    void reset([[maybe_unused]] std::lock_guard<std::mutex> & cache_lock) override
+    {
+        queue.clear();
+        cells.clear();
+        current_size = 0;
+    }
+
+    void remove(const Key & key, [[maybe_unused]] std::lock_guard<std::mutex> & cache_lock) override
+    {
+        auto it = cells.find(key);
+        if (it == cells.end())
+            return;
+        auto & cell = it->second;
+        current_size -= cell.size;
+        queue.erase(cell.queue_iterator);
+        cells.erase(it);
+    }
+
+    MappedPtr get(const Key & key, [[maybe_unused]] std::lock_guard<std::mutex> & cache_lock) override
+    {
+        auto it = cells.find(key);
+        if (it == cells.end())
+        {
+            return MappedPtr();
+        }
+
+        Cell & cell = it->second;
+
+        /// Move the key to the end of the queue. The iterator remains valid.
+        queue.splice(queue.end(), queue, cell.queue_iterator);
+
+        return cell.value;
+    }
+
+    void set(const Key & key, const MappedPtr & mapped, [[maybe_unused]] std::lock_guard<std::mutex> & cache_lock) override
+    {
+        auto [it, inserted] = cells.emplace(std::piecewise_construct,
+            std::forward_as_tuple(key),
+            std::forward_as_tuple());
+
+        Cell & cell = it->second;
+
+        if (inserted)
+        {
+            try
+            {
+                cell.queue_iterator = queue.insert(queue.end(), key);
+            }
+            catch (...)
+            {
+                cells.erase(it);
+                throw;
+            }
+        }
+        else
+        {
+            current_size -= cell.size;
+            queue.splice(queue.end(), queue, cell.queue_iterator);
+        }
+
+        cell.value = mapped;
+        cell.size = cell.value ? weight_function(*cell.value) : 0;
+        current_size += cell.size;
+
+        removeOverflow();
+    }
+
+protected:
+    using LRUQueue = std::list<Key>;
+    using LRUQueueIterator = typename LRUQueue::iterator;
+
+    LRUQueue queue;
+
+    struct Cell
+    {
+        MappedPtr value;
+        size_t size;
+        LRUQueueIterator queue_iterator;
+    };
+
+    using Cells = std::unordered_map<Key, Cell, HashFunction>;
+
+    Cells cells;
+
+    /// Total weight of values.
+    size_t current_size = 0;
+    const size_t max_size;
+    const size_t max_elements_size;
+
+    WeightFunction weight_function;
+
+    void removeOverflow()
```

Comment:
btw may be take a look at https://github.com/ClickHouse/ClickHouse/pull/38048, there is also an attempt to make an abstract cache policy class, but it is different from your as it does not make a whole implementation of methods like `removeOverflow` (in your case) implemented in each policy class, but changes only small parts. I like his approach more, may be try to do something similar here? what do you think?

Reply:
If I correctly understand you suggested adding a function that returns an iterator (or something else) on the element with the lowest priority and then deletes it. This approach doesn't simplify code here because we should also move elements between queues and maintain the invariant for max size of all cache and protected part. So after each operation, we should use a  complex solution to work with policy in the right way and realization of this will work exactly like now but with additional abstraction which is not really necessary in my opinion

Reply:
Ok, not necessary I agree, just a suggestion as in the mentioned pr it allowed to avoid a lot of copy paste, but here the logic is more simple so not clear if it would be better or not.